packages/react-devtools-shared/src/__tests__/profilingCache-test.js

```diff
@@ -749,4 +754,671 @@ describe('ProfilingCache', () => {
+        </>,
+        container,
+      ),
+    );
+
+    utils.act(() => store.profilerStore.stopProfiling());
+
+    function getPerfInsightSnapshot(index: number) {
+      const rootID = store.roots[0];
+      const commitData = store.profilerStore.getCommitData(rootID, index);
+      return commitData.perfInsights;
+    }
+
+    // First mount, expect everything to be false.
+    expect(getPerfInsightSnapshot(0)).toMatchInlineSnapshot(`
+      Map {
+        2 => Object {
+          "didContextChange": false,
+          "didContextDeepChange": false,
+          "didHooksChange": false,
+          "didPropsChange": false,
+          "didPropsDeepChange": false,
+          "didStateChange": false,
+          "didStateDeepChange": false,
+          "isFirstMount": true,
+          "nonMemoizedProps": null,
+        },
+        3 => Object {
+          "didContextChange": false,
+          "didContextDeepChange": false,
+          "didHooksChange": false,
+          "didPropsChange": false,
+          "didPropsDeepChange": false,
+          "didStateChange": false,
+          "didStateDeepChange": false,
+          "isFirstMount": true,
+          "nonMemoizedProps": null,
+        },
+        4 => Object {
+          "didContextChange": false,
+          "didContextDeepChange": false,
+          "didHooksChange": false,
+          "didPropsChange": false,
+          "didPropsDeepChange": false,
+          "didStateChange": false,
+          "didStateDeepChange": false,
+          "isFirstMount": true,
+          "nonMemoizedProps": null,
+        },
+        5 => Object {
+          "didContextChange": false,
+          "didContextDeepChange": false,
+          "didHooksChange": false,
+          "didPropsChange": false,
+          "didPropsDeepChange": false,
+          "didStateChange": false,
+          "didStateDeepChange": false,
+          "isFirstMount": true,
+          "nonMemoizedProps": null,
+        },
+        6 => Object {
+          "didContextChange": false,
+          "didContextDeepChange": false,
+          "didHooksChange": false,
+          "didPropsChange": false,
+          "didPropsDeepChange": false,
+          "didStateChange": false,
+          "didStateDeepChange": false,
+          "isFirstMount": true,
+          "nonMemoizedProps": null,
+        },
+      }
+    `);
+
+    // Second render. Only non memoized component updates didPropsChange is false.
+    expect(getPerfInsightSnapshot(1)).toMatchInlineSnapshot(`
+      Map {
+        2 => Object {
+          "didContextChange": false,
+          "didContextDeepChange": false,
+          "didHooksChange": false,
+          "didPropsChange": false,
+          "didPropsDeepChange": false,
+          "didStateChange": false,
+          "didStateDeepChange": false,
+          "isFirstMount": false,
+          "nonMemoizedProps": null,
+        },
+      }
+    `);
+
+    // Third render. We increment count and expect props to deep change.
```

Comment:
I'm confused. A number prop can't "deep change". Code comments above agree with this:
```js
    // Third render has shallow changes only.
    utils.act(() =>
      legacyRender(
        <>
          <ComponentNeedsMemo count={2} />
          <ComponentShallowMemo count={2} />
          <ComponentNeedsDeepMemo style={style} />
          <ComponentDeepMemo style={style} />
        </>,
        container,
      ),
    );
```

So why do perf insights below show `didPropsDeepChange: true`?

Reply:
Maybe I can rename all the `DeepChange` to `DeeplyChange`?

Reply:
>> A number prop can't "deep change". Code comments above agree with this:
>
> In this case, the props did change, even with a deep equality check. Hence, did the props deeply change? Yes.

I still don't understand.

A primitive value like a number _can't "deeply" change_. This concept only makes sense for objects or arrays.

DevTools should never tell a user that a number or a string "deeply changed". It would confuse them.

Reply:
We aren't saying the a string or number deeply change, rather the props *object* deeply changed. 

I think its fair to say `{foo: 'bar'}` and `{foo: 'baz'}` are deeply differently. It's not implying `'bar'` and `'baz'` are deeply different.

Reply:
> rather the props object deeply changed.

I'm not sure that's really what it means?

Looking at [`changedPropKeys`](https://github.com/facebook/react/pull/22723/files?w=1#diff-d4e1c1c8b2c1c6fe180fe75e2dfb431fa7da9612dddef31dc709ddc5e7973e7eR1285-R1288) for example:
```js
const changedPropKeys = getChangedKeys(
  prevFiber.memoizedProps,
  nextFiber.memoizedProps,
);
```

This calls `getChangedKeys` which just loops and returns a list of _keys_ that changed:
https://github.com/facebook/react/blob/afbc2d08f4dabf00222e9561344b65e51d6ed190/packages/react-devtools-shared/src/backend/renderer.js#L1494-L1501

Okay. This makes sense. Any keys that changed are "changed keys".

My issue (or my confusion) is with `deepChangedPropKeys`.

First a simple case:
```js
<Component foo={object} />
<Component foo={{...object}} />
```

Only `changedPropKeys` contains `foo`. That makes sense, since the `foo` object's top level props are all equal. In this case, `deepChangedPropKeys` means "nothing inside of this object changed".

Next a confusing case:
```js
<Component foo={1} />
<Component foo={2} />
```
Both `changedPropKeys` and `deepChangedPropKeys` contain "foo". The value `foo` is a number so it _can't deeply change_. This is fundamentally confusing, given the naming.

Next an incorrect case:
```js
<Component foo={{nested: object}} />
<Component foo={{nested: {...object}}} />
```
This will incorrect flag the `foo` prop as being _deeply equal_ too, when it's not.

Maybe a rename of these properties is needed to make this clearer. I think the _intention_ here is to be able to identify "this thing would not have re-rendered at all if you memoized b'c props and state objects were referentially equal"? If so, maybe we can use a clearer property to reflect this.

Also, it's not clear to me that "hooks" and "context" deep vs not-deep distinctions mean the same thing. If not, using a similar name makes things even more confusing.

Reply:
You know what they say are the hardest things in programming..

Reply:
We talked out of band a bit and the tension seems to be coming from the fact that some libraries (e.g. lodash `isEqual`) support comparing primitive values for "deep equality" – which I think makes sense. It would be weird for `isEqual(2, 2)` to return false. At the same time, it's weird to refer to a primitive value (string, number, boolean, etc) as having "deep" properties.

What if we borrowed a term from another language like "structural equality" – comparing the content stored within the values rather than the memory space the variables point to.

I think this means the same thing as Lodash's method, but it's less semantically confusing to me than "deep", at least initially.

Reply:
Couple of suggestions:
`deepChangedPropKeys` -> 
- `nonIdenticalPropKeys` / `nonIdenticalProps`
- `nonEquivalentPropKeys` /  `nonEquivalentProps`
- `nonEquivalentValuePropKeys` / `nonEquivalentValueProps`
- `propsKeysNotEqualOrDeepEqual` / `propsNotEqualOrDeepEqual`

Reply:
I think I like the term "equivalent" better, but I was thinking maybe we should take a step back and look at this.

Given the following before and after values, here's how I think about them:

|  | Value 1 | Value 2 | Referentially equal | Structurally Equal |
| --- | --- | --- | --- | --- |
| A | `1` | `2` | No | N/A |
| B | `{ foo: 1 }` | `{ foo: 1 }` | No | Yes |
| C | `{ foo: 1 }` | `{ foo: 2 }` | No | No |
| D | `STYLE_CONST` | `STYLE_CONST` | Yes | Yes |

I think the core of my concern is case A. I find any terminology that describes A as being "deeply changed" (or anything similar) to be confusing. (Structural equality isn't relevant because a number doesn't have any "structure". It's a primitive.) So this is the case I'm trying to find an alternate for

I think the core of your concern is avoiding ambiguity betwen cases A and B. You want a way to know when to say "_A changed and no type of memoization could have prevented this_" vs "_B changed but if you were using a different type of memoization check, like Lodash `isEqual`, you could have avoided a re-render_".

I wonder how we can address both?

Maybe we instead of describing the value as "_is it equal?_" and "_is it deeply equal?_", we could try destribing it as "_is it equal?_" and "_would a structural equality check allow this to be memoized?_" I don't know what the specific name for this second thing would be, but it seems like it would avoid both the confusing naming case and the ambiguity case?

|  | Value 1 | Value 2 | Referentially equal | Could still be memoized? |
| --- | --- | --- | --- | --- |
| A | `1` | `2` | No | No |
| B | `{ foo: 1 }` | `{ foo: 1 }` | No | Yes |
| C | `{ foo: 1 }` | `{ foo: 2 }` | No | No |
| C | `STYLE_CONST` | `STYLE_CONST` | Yes | Yes |

Reply:
The first one that poops into mind is `nonMemoizablePropKeys`, but its misleading because it _could_ be memoized, but not in this scenario (with these specific values).

`propsKeysNotEqualOrDeepEqual` / `propsNotEqualOrDeepEqual` seem to satisfy the table?

`keysNotBenefitingFromMemoization` is another idea.

Reply:
Ooo, what if we modeled the data a little differently?

Given the same data as above:
|  | Value 1 | Value 2 |
| --- | --- | --- |
| A | `1` | `2` |
| B | `{ foo: 1 }` | `{ foo: 1 }` |
| C | `{ foo: 1 }` | `{ foo: 2 }` |
| D | `STYLE_CONST` | `STYLE_CONST` |

What if we represented it as an object with _changed_ prop names as keys?
```js
changedProps: {
  A: { couldBeMemoized: false },
  B: { couldBeMemoized: true },
  C: { couldBeMemoized: false } 
}
```
We could potentially expand this format in the future, e.g.
```js
changedProps: {
  A: { couldBeMemoized: false, from: "1", to: "2" },
  B: { couldBeMemoized: true, from: "{ foo: 1 }", to: "{ foo: 1 }" },
  C: { couldBeMemoized: false, from: "{ foo: 1 }", to: "{ foo: 2 }" },
}
```
Not that we'd necessarily want to pay the cost of serializing before and after values while profiling, but this _has been requested_ from time to time.

Reply:
I'm honestly okay with any name change to get this PR in; but I do think `memoizable: false` is a bit misleading. What we really know is that this value wasn't equal or structurally equal, not whether is is _not_ memoizable.

We could just go back to the ideal of equal, since lodash's method is called `isEqual` and not `isDeepEqual`.
`nonEqualPropKeys`.

If your fine with `changedProps: {[prop]: {memoizable: boolean}` I'm more than happy to make the change.

Reply:
> What we really know is that this value wasn't equal or structurally equal, not whether is is not memoizable.

That's fair. Maybe "memoizable" is too terse. What I was going for was– "could a render be avoided using memoization with a structural comparison?"

Revisiting this today, I like using an object with keys for other reasons besides this naming issue. It allows us to add in other things (like from/to) if we choose to.

I dunno. Maybe...
```js
changedProps: {
  A: { isStructurallyEqual: false },
  B: { isStructurallyEqual: true },
  C: { isStructurallyEqual: false },
},
changedState: {
  // ...
},
changedContext: {
  // ...
}
```

Reply:
I'm okay with implementing that, but `isStructurallyEqual` would be true for primitives. We could be verbose and say `isEqualOrStructurallyEqual` or something else?

Reply:
No, `isStructurallyEqual` would never be true for primitives. The only keys/props in this object are ones that _changed_.

Reply:
@bvaughn I updated the variable naming. Instead of checking for deep equality, I'm inferring if props/state/hook can benefit from memoization which is the main goal.

Instead of `didPropsDeeplyChange` -> `canPropsBeMemoized`

Reply:
I dig it.

Reply:
I dig the "can be memoized" name.